lib/stub_ui/lib/src/ui/text.dart

```diff
@@ -1011,6 +1011,185 @@ enum BoxWidthStyle {
+  final double descent;
+
+  /// The rise from the [baseline] as calculated from the font and style for this line
+  /// ignoring the text scale factor.
+  ///
+  /// The [unscaledAscent] is provided as a positive value, even though it is typically
+  /// defined in fonts as negative. This is to ensure the signage of operations with
+  /// these metrics directly reflects the intended signage of the value. For example,
+  /// the y-coordinate of the top edge of the line is 'baseline - ascent`.
+  final double unscaledAscent;
+
+  /// Total height of the line from the top edge to the bottom edge.
+  ///
+  /// This is equivalent to `ascent + descent`
+  final double height;
+
+  /// Width of the line.
+  final double width;
+
+  /// The x-coordinate of left edge of the line.
+  ///
+  /// The right edge can be obtained with `left + width`.
+  final double left;
+
+  /// The y position of the baseline for this line from the top of the paragraph.
+  final double baseline;
+
+  /// Zero indexed line number.
+  ///
+  /// For example, the first line is line 0, second line is line 1.
+  final int lineNumber;
+
+  /// Mapping between text index ranges and the RunMetrics associated with
+  /// them.
+  ///
+  /// Each entry will be keyed under the ending index the run applies to. Each
+  /// [RunMetrics] entry corresponds to the text covered by the range of
+  /// `[previous key, current key)`. For a given index `i`, the [RunMetrics]
+  /// that applies to it is the largest key such that `i < key`. For the first
+  /// key, the [startIndex] acts as the previous key.
+  ///
+  /// The metrics here are pre-layout values and are the base font metrics we
+  /// compute the rest of the paragraph from. These metrics do not account for
+  /// modifiers such as [StrutStyle] or `textScaleFactor`.
+  final Map<int, RunMetrics> runMetrics;
+}
+
+/// Data class that tracks the raw font metrics for a particular text buffer index.
+class RunMetrics {
+  @pragma('vm:entry-point')
+  RunMetrics._(
+    this.top,
+    this.ascent,
+    this.descent,
+    this.bottom,
+    this.leading,
+    this.avgCharWidth,
+    this.maxCharWidth,
+    this.xMin,
+    this.xMax,
+    this.xHeight,
+    this.capHeight,
+    this.underlineThickness,
+    this.underlinePosition,
+    this.strikeoutThickness,
+    this.strikeoutPosition,
+  );
+
+  // TODO(garyq): Include the TextStyle here too.
+
+  /// Extent above baseline.
+  final double top;
+  /// Distance to reserve above baseline.
+  ///
+  /// Since this is the raw metric from the font, [ascent] is usually negative.
+  final double ascent;
+  /// Distance to reserve below baseline.
+  final double descent;
+  /// Extent below baseline.
+  final double bottom;
+  /// Additional distance to add between lines.
+  final double leading;
+  /// Average character width.
+  final double avgCharWidth;
+  /// Maximum character width.
+  final double maxCharWidth;
+  /// Minimum x.
```

Comment:
What's x?

Reply:
That is a good question, I am not totally sure myself, but I will find out. This is one of the metrics from SkFontMetrics, but we don't seem to use it anywhere. It is provided to us nonetheless.

There is some ambiguous naming going on where X refers to x-coordinate as well as x the ascii letter. I can try to clarify this in our API.